Review Request: Demo Additions on top of PRs #43 and #45

awwad opened the pull request:
This is code written to correct and add to PRs #43 and #45. This already-merged code should probably have been reviewed, so this is a PR that will not be merged, for the purpose of review.

Among the changes, I cut out the final attack (which has to be replaced - see commit messages), made repository recovery minimal (rather than nuking the repo, do key revocation and corrupted target replacement), switched from revoking just Targets to all likely-compromised keys (Timestamp, Snapshot, and Targets in this scenario), stripped an unnecessary bit, and added to some documentation.

awwad committed:
DOC: Add links to some sections of the demo setup

awwad committed:
DOC: Make command context (prompt, shell, etc.) clearer in README

by adding explicit instructions to open a Python shell, open a new
window, etc.

awwad committed:
DOC: Markdown ```python for Python code sections

awwad committed:
DOC: Strip trailing whitespace in README

awwad committed:
DOC: Restructure heading levels in README, strip italics

and rename a few headings.

awwad committed:
DOC: Add table of contents to README

awwad committed:
DOC: clarify confusing sentence in README

awwad committed:
DOC: Remove erroneous content in README

awwad committed:
DOC: Add some missed Python prompts to README

awwad committed:
DOC: Number windows in early parts of the README

and also remove an instance of trailing whitespace.

awwad committed:
DEMO: Clarify replay attack banner slightly

awwad committed:
DOC: Clarify MITM attack instructions in README

awwad committed:
DOC: Clarify replay attack instructions in README

awwad committed:
DOC: Substantially simplify the later demo attacks in README

It is not necessary to add an additional target, new_firmware.img,
as we can use the existing one, firmware.img, for the same attack.

awwad committed:
DOC: Clarify docstring on number of keys used in Director

awwad committed:
DOC: Remove duplicate instructions in docstring in demo Director

for rollback attacks.

awwad committed:
Python3 compatibility fixes to the PR

Unicode encoding issues. Also, an old line comes back to haunt
us and so has been stripped (from io import open).

awwad committed:
Revamp recovery code for demo Director and demo Image Repo

Instead of just replacing the Targets keys, we're going to replace
all potentially-compromised keys: Timestamp, Snapshot, and Targets.

awwad committed:
DOC: Revamp end of demo: recovery and final description

Instead of exiting the servers, we repair in place, doing only what
is necessary (key revocations, replacement of images).

Also added description about what's happening and why.

Also removed the last attack, which seemed to claim to demonstrate
what it was not demonstrating: it seemed to suggest that it was
showing that the revoked keys were no longer trusted, but the
attack did not employ those keys. A further attack should be added
that does. Issue #63 has been added to document this.

awwad committed:
Demo Director add target script no longer spams every vehicle

write_to_live() was used without identifying the single
VIN to update, so each time a single vehicle was updated using
this function in the demo, it write new metadata for each
vehicle.

awwad committed:
Merge branch 'vladimir-v-diaz-package_attacks' into develop

awwad committed:
Cut down on some confusing demo spam

relating to not-yet-used C CAN interface. Also add to a few comments.

awwad committed:
DOC: Correcting typos and numbering in README

vladimir-v-diaz commented on README.md:
```diff
@@ -43,36 +55,41 @@ If you want the demo to play notification sounds you need one of the following a
 These instructions start a demonstration version of an OEM's or Supplier's main repository
 for software, hosting images and the metadata Uptane requires.
 
+From within the root `uptane/` directory of the repository, open an interactive
```
"... directory of the **repository**"
The word repository here can mean multiple things, a GitHub repository, an Uptane/TUF repository, etc.
Consider using another word to avoid confusion, like "source files/code".  Or you can elaborate by saying the `uptane` directory that contains the setup.py file.

vladimir-v-diaz commented on README.md:
```diff
@@ -128,24 +151,26 @@ The Primary's update_cycle() call:
 
-If you wish to run the demo with multiple vehicles (one Primary each), you can open a new
+If you wish to run the demo with multiple vehicles (one Primary each), you can open a Python shell in a new terminal
 window for each vehicle's Primary and provide a unique VIN and ECU for each of them. Find the port that is chosen in the Primary's initialization and make note of it so that it can be provided to any Secondaries you set up in a moment (e.g. "Primary will now listen on port 30702")
```
Find the port that is chosen in the Primary's initialization message (e.g., "Primary will not listen on port 30702") and make note of it so that it can be provided to any Secondaries you set up in a moment.

It might be unclear which function call initializes the primary.  Is it dp.clean_slate()?  It might not be obvious to new users from this name alone.

awwad replied:
True, clarifying.

vladimir-v-diaz commented on README.md:
```diff
@@ -128,24 +151,26 @@ The Primary's update_cycle() call:
 
-If you wish to run the demo with multiple vehicles (one Primary each), you can open a new
+If you wish to run the demo with multiple vehicles (one Primary each), you can open a Python shell in a new terminal
 window for each vehicle's Primary and provide a unique VIN and ECU for each of them. Find the port that is chosen in the Primary's initialization and make note of it so that it can be provided to any Secondaries you set up in a moment (e.g. "Primary will now listen on port 30702")
```
Should ECU be ECU_SERIAL?

awwad replied:
Yes.

vladimir-v-diaz commented on README.md:
```diff
@@ -226,16 +251,17 @@ Since this file is not on (and validated by) the Image Repository, the Primary w
-normal state (undoing what the attack did) by running the following in the Director's window:
+To resume experimenting with the repositories, run this script to put the
+repository back in a normal state (undoing what the attack did) by running the
+following in the **Director's** window:
```
Include the WINDOW # here, like you do for the others.

vladimir-v-diaz commented on README.md:
```diff
@@ -7,8 +7,20 @@ Please note that extensive documentation on design can be found in the following
+Below are the instructions on use of the Uptane demonstration and reference
+implementation code, divided into these sections:
+
+* [0: Installation](#0-installation)
```
Should the Table of Contents also include subheadings?  An outline of the demo's workflow and attacks might help the reader.

awwad replied:
Sure. Adding.

vladimir-v-diaz commented on README.md:
````diff
@@ -335,77 +380,56 @@ repository key and signs for a malicious image?  Is the client able to block a c
->>> dd.add_target_and_write_to_live(filename='new_firmware.img', file_content='new firmware image',
-    vin='111', ecu_serial='22222')
-```
+For this attack, we'll modify that to include malicious content.
````
"... we'll modify _that_ ..."  Instead of that, refer to the actual file.

vladimir-v-diaz commented on demo/demo_image_repo.py:
```diff
@@ -26,7 +26,6 @@
 """
 from __future__ import print_function
 from __future__ import unicode_literals
-from io import open
```
I think there are other modules that import io.open(): demo_primary.py and demo_secondary.py
Should we open an issue for this?

awwad replied:
I'll try removing them locally and run tests and the demo and if it works, their removal will come in a future commit.

awwad committed:
DOC: Respond to Vlad's review comments with some minor changes

- Add subheadings to table of contents for attack demos
- (Also shortened subheadings for this purpose)
- Clarify comment on what directory to open shells from

I also fixed a minor typo and dereferenced a pronoun.

awwad committed:
DOC: Add blurb on compromised devices to README

awwad commented:
LMK if/when it's good/closable, so it doesn't accumulate anything added to develop.

vladimir-v-diaz commented on README.md:
```diff
@@ -404,7 +410,7 @@ us to download a file that does  does not exactly match the Image Repository met
 
 
-### 3.5: Compromise the Image repository to also serve the arbitrary package
+### 3.5: Compromise Both Repositories Simultaneously to Serve Arbitrary Package
```
This section only includes instructions to compromise the image repository, but it does build upon the previous section so that both repositories are compromised.

vladimir-v-diaz commented:
LGTM.  I just had one minor comment.

awwad commented:
Will address remaining comment and some other clarifying modifications to the README in another pull request. Closing this (:
